svcdef/consolidate_http.go

```diff
@@ -0,0 +1,132 @@
+		lex := svcparse.NewSvcLexer(pfile)
+		protosvc, err := svcparse.ParseService(lex)
+		if err != nil {
+			if strings.Contains("'options' or", err.Error()) {
+				log.Warnf("Parser found rpc method which lacks HTTP " +
+					"annotations; this is allowed, but will result in HTTP " +
+					"transport not being generated.")
+			} else {
+				return errors.Wrap(err, "error while parsing http options for the service definition")
+			}
+		}
+		assembleHTTPParams(sd.Service, protosvc)
+	}
+	return nil
+}
+
+// assembleHTTPParams will use the output of the service parser to create
+// HTTPParams for each service RequestType field indicating that parameters
+// location, and the field to which it refers.
+func assembleHTTPParams(svc *Service, httpsvc *svcparse.Service) error {
+	getMethNamed := func(name string) *ServiceMethod {
+		for _, m := range svc.Methods {
+			if m.Name == name {
+				return m
+			}
+		}
+		return nil
+	}
+
+	// Since this functionality must be repeated for each HTTPBinding on each
+	// ServiceMethod, and since this logic would be a bit hard to make sense of
+	// nested in a for loop, it's been broken out into this function. Since
+	// it's only used within this function, and is purely to make this logic
+	// more readable, I chose to define it within this function.
+	createParams := func(meth *ServiceMethod, parsedbind *svcparse.HTTPBinding) {
+		msg := meth.RequestType.Message
+		bind := HTTPBinding{}
+		bind.Verb, bind.Path = getVerb(parsedbind)
+
+		params := make([]*HTTPParameter, 0)
+		for _, field := range msg.Fields {
+			new_param := &HTTPParameter{}
+			new_param.Field = field
+			new_param.Location = paramLocation(field, parsedbind)
+			params = append(params, new_param)
+		}
+		bind.Params = params
+		meth.Bindings = append(meth.Bindings, &bind)
+	}
+
+	// Iterate through every HTTPBinding on every ServiceMethod, and create the
+	// HTTPParameters for that HTTPBinding.
+	for _, hm := range httpsvc.Methods {
+		m := getMethNamed(hm.Name)
+		if m == nil {
+			return errors.New(fmt.Sprintf("Could not find service method named %q", hm.Name))
+		}
+		for _, hbind := range hm.HTTPBindings {
+			createParams(m, hbind)
+		}
+	}
+	return nil
+}
+
+// getVerb returns the verb of a svcparse.HTTPBinding. If the binding does not
+// contain a field with a verb, returns two empty strings. Currently doesn't
+// support "custom" verbs.
+func getVerb(binding *svcparse.HTTPBinding) (verb string, path string) {
+	for _, field := range binding.Fields {
+		switch field.Kind {
+		case "get", "put", "post", "delete", "patch":
+			return field.Kind, field.Value
+		}
+	}
+	return "", ""
+}
+
+// paramLocation returns the location that a field would be found according to
+// the rules of a given HTTPBinding.
+func paramLocation(field *Field, binding *svcparse.HTTPBinding) string {
+	path_params := getPathParams(binding)
+	for _, path_param := range path_params {
+		if strings.Split(path_param, ".")[0] == field.Name {
```

Comment:
That behavior is an undocumented hack, as it turns out...

I should point out that most of the `consolidateHTTP` code is taken straight from `doctree`, so this code was written relatively early in the life of truss.

The point of this line of code is to paper over a not-often-used feature that I didn't know how to implement, namely specifying the location of sub-parameters of messages in the path. Here's an example:

```
message MsgA {
    int a = 1;
    MsgB b = 2;
}
message MessageB {
    int Foo = 1;
}
service Exmpl {
  rpc GetA (MsgA) returns (MsgA) {
    option (google.api.http) = {
      get: "/{b.Foo}"
    };
  }
}
```

The above is totally valid, and *eventually* we should support this. But for right now, I've not implemented this. Instead, the gross hack that I have above would make svcdef thing that just `b` was embedded in the path.

I've added a comment noting this incompatibility.